### PR TITLE
CY-2285 Workflow execution model docs

### PR DESCRIPTION
**Pull request by tehasdf**

### CODE REVIEW

**Comment by tehasdf** [comment]
preview is https://docs.cloudify.co/staging/execution-model-docs-1/developer/execution_model/

**Comment by tehasdf** [comment]
@ofercloudify this is ready

File: content/developer/execution_model.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        }
+
+1. The management worker receives the "start-workflow" message from rabbitmq, and runs a "dispatcher" subprocess to handle it.
+1. The subprocess loads and executes the workflow function. Most workflow functions create a tasks graph and execute it.
+1. The subprocess periodically checks execution state to react to the state changing to CANCELLED.
+1. The workflow function might store the tasks graph and the operations, to allow resuming.
+1. For every operation in the tasks graph, the dispatcher process sends "start-operation" messages:
+    - the operation is transitioned to the SENDING state, and then SENT
+    - the message is sent to the exchange `<agent_name>` for host_agent tasks, or `cloudify.management` for central_deployment_agent tasks
+    - the dispatcher process starts listening for the task result on the `<agent_name>_response_<task_id>`.
```
Comment (geokala) [inline content/developer/execution_model.md]: Is this the same (with agent_name = cloudify.management) for central_deployment_agent?

File: content/developer/execution_model.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    The "start-operation" message contains at least the following fields:
+
+        {
+            'id': task_id,
+            'cloudify_task': {
+              'kwargs': dict containing the operation parameters, including a special parameter '__cloudify_context'
```
Comment (geokala) [inline content/developer/execution_model.md]: Is it deliberately bold from the start of __cloudify_context through the rest of the file, or is this a bug on my side?

File: content/developer/execution_model.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            }
+        }
+
+1. The target agent or mgmtworker receives the message and starts an operation subprocess to execute the task:
+    - operation message is acked
+    - task is transitioned to STARTED
+    - operation function is executed
+1. The agent finishes executing the task
+    - task is transitioned to a terminal state (SUCCEEDED, FAILED, or RESCHEDULED)
+    - a "operation-response" message is sent.
```
Comment (geokala) [inline content/developer/execution_model.md]: *an "operation-response"...
(or, in keeping with the structure of the previous list item, just - "operation-response"...)

File: content/developer/execution_model.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        {
+          "ok": boolean,
+          "result": operation result value
+        }
+
+    And optionally an "error" field.
+1. The dispatcher receives the response
+    - the dispatcher deletes the task response queue
+1. After all tasks have been executed, the dispatcher finishes executing the workflow
+    - execution state is changed to TERMINATED or FAILED
```
Comment (geokala) [inline content/developer/execution_model.md]: Maybe note that terminated is considered to be completed (or whatever we call it) elsewhere in the system to make it more clear that TERMINATED == good.

File: content/developer/execution_model.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+              }
+          }
+        }
+
+1. Every agent receives the message, and kills every running operation subprocess of the cancelled execution by sending SIGTERM, and 5 seconds later, SIGKILL
+
+
+## Resuming an execution
+
+To kill-cancel an execution, the user sends a POST request with the parameter "action" set to "resume" or "force-resume".
```
Comment (geokala) [inline content/developer/execution_model.md]: To resume an execution...

File: content/developer/execution_model.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+Resuming STARTED executions is allowed to help restore "stuck" executions; to force-resume a STARTED execution, it should be cancelled first (if it is truly "stuck", it will have to be kill-cancelled).
+
+1. The restservice updates the execution object with:
+    - a new execution token
+    - ended_at date set to null
+    - status set to STARTED
+1. Update stored operations: operations in state RESCHEDULED or FAILED are set to PENDING, and their current_retries count is reset to 0.
+1. In case of a force-resume, also do the previous step for operations in state STARTED, SENT, or SENDING.
+1. Start running the execution as in the "Running an execution" section, with the exception that:
+    - where the management worker sends "start-operation" messages, if the operation is already in a SENT or STARTED state, the message is not sent. Instead, the management worker proceeds to the next step, and waits for a response.
```
Comment (geokala) [inline content/developer/execution_model.md]: I assume there's more to come, given the 'proceeds to the next step' line?

File: content/developer/execution_model.md (hunk 2 above)
Comment (tehasdf) [reply content/developer/execution_model.md]: github renders this slightly differently than it's going to be rendered on the docs side, please see the preview at https://docs.cloudify.co/staging/execution-model-docs-1/developer/execution_model/
(not bold there, and tbh no idea why is it bold here)

File: content/developer/execution_model.md (hunk 1 above)
Comment (tehasdf) [reply content/developer/execution_model.md]: reworded the previous line

File: content/developer/execution_model.md (hunk 3 above)
Comment (tehasdf) [reply content/developer/execution_model.md]: `an` flows better for me here

File: content/developer/execution_model.md (hunk 6 above)
Comment (tehasdf) [reply content/developer/execution_model.md]: no, note that this is under a `Start running the execution as in the "Running an execution" section, with the exception that`. So, proceeds to the next step, of that other section. Reworded a bit

**Comment by tehasdf** [comment]
@ofercloudify please use `Squash and merge` when merging

**Pull request merged**